Add readme and initial values (#1)
OpenStreetMapSpeeds/schema: README.md

# OpenStreetMapSpeeds Schema

> This repository houses both a discription and outline of the JSON schema for the data artifact the conflation repository produces as well as the latest version of the artifact itself.

Reviewer: typo in `description`

Author: fixed

## Schema

> Broadly the schema is a collection of nested JSON objects. At the top level we have an array of objects where each object represents a geographic region. The region may be specified using the optional [iso3166-1 country](https://en.wikipedia.org/wiki/ISO_3166-1) and [iso3166-2 principal subdivision](https://en.wikipedia.org/wiki/ISO_3166-2) codes. These geographies are treated as a heirarchy to allow varying degrees of specificity. For example one can specify a global set of speeds (ie applies to the whole world) by simply omitting both the `iso3166-1` and `iso3166-2` keys in a given object. To specify speeds accross an entire country one need only populate the `iso3166-1` key and omit the `iso3166-2` key. For full granularity one must include both `iso3166-1` and `iso3166-2`.

Reviewer 2: ```suggestion
Broadly the schema is a collection of nested JSON objects. At the top level we have an array of objects where each object represents a geographic region. The region may be specified using the optional [iso3166-1 country](https://en.wikipedia.org/wiki/ISO_3166-1) and [iso3166-2 principal subdivision](https://en.wikipedia.org/wiki/ISO_3166-2) codes. These geographies are treated as a hierarchy to allow varying degrees of specificity. For example one can specify a global set of speeds (i.e. applies to the whole world) by simply omitting both the `iso3166-1` and `iso3166-2` keys in a given object. To specify speeds across an entire country one needs to only populate the `iso3166-1` key and omit the `iso3166-2` key. For full granularity one must include both `iso3166-1` and `iso3166-2`.
```

Within a given object there are 3 divisions based on assumed changes in road network congestion, these are `rural`, `suburban` and `urban`. We deliberately avoid a one size fits all defintion of these categorizations as they may differ from system to system. See the [conflation](https://github.com/OpenStreetMapSpeeds/conflation) repo for more information about where the data comes from and how it is used to populate these divisions.

Within each density/populational division we further classify by road types which should be somewhat familiar to those familiar with OSM. There are really only two types of information we use to do this classification, the functional road class (FRC or `highway` tag in OSM) and the form of way (FoW may come from many different tag values in OSM). The `highway` tag values that we assign speeds for are in descending importance: `motorway`, `trunk`, `primary`, `secondary`, `tertiary`, `unclassified`, `residential` and `service`.

> We'll start by describing those values who only rely on their form of way, ie `driveway`, `alley`, `parking_aisle` and `drive-through`. These are all tagged in OSM as `highway=service` and then `service=*` where `*` is one of `driveway`, `alley`, `parking_aisle` or `drive-through`. Because they have only one value for their `highway` tag they only apply to that 1 functional road class (the lowest) and so they have only a single value for speed.

Reviewer 2: ```suggestion
We'll start by describing those values who only rely on their form of way, ie `driveway`, `alley`, `parking_aisle` and `drive-through`. These are all tagged in OSM as `highway=service` and then `service=*` where `*` is one of `driveway`, `alley`, `parking_aisle` or `drive-through`. Because they have only one value for their `highway` tag they only apply to that 1 functional road class (the lowest, i.e. `service`) and so they have only a single value for speed.
```

> We also differentiate roundabouts which are tagged as `junction=roundabout` on OSM ways. Since they may occur with any `highway` tag from above we have an array for their speeds where the first speed in the array corresponds to the `motorway` speed and the last speed in the array corresponds to the `service` speed.

Reviewer 2: ```suggestion
We also differentiate roundabouts which are tagged as `junction=roundabout` on OSM ways. Since they may occur with any `highway` tag from above we have an array for their speeds where the first speed in the array corresponds to the `motorway` speed and the last speed in the `roundabout` array corresponds to the `service` speed.
```

For the `highway=*_link` tags we have two classifications. Those that appear with any kind of signage (eg. `destination=*`) which we call `link_exiting` or those without signage which we call `link_turning`. Note that we only support an array of speeds for the 5 most important `highway=*_link` tags ie. `motorway_link`, `trunk_link`, `primary_link`, `secondary_link` and `tertiary_link`.

> Finally we have all other ways which do not have a designated or implied form of way. For these we simply supply a speed per `highway` tag as mentioned above, where the first speed is for `highway=motorway` and the last speed is for `highway=service`.

Reviewer 2: ```suggestion
Finally we have all other ways which do not have a designated or implied form of way in the `way` array. For these we simply supply a speed per `highway` tag as mentioned above, where the first speed is for `highway=motorway` and the last speed is for `highway=service`.
```

Author: ill reword this one!

Author: done

> The schema, since it uses arrays does not allow for the omission of values however one could use `0` or `null` to signal no data for a particular categorization.

Reviewer 2: guess we need to update valhalla then soon?:)

```suggestion
The schema, since it uses arrays, does not allow for the omission of values however one could use `0` or `null` to signal no data for a particular categorization.
```

Author: yep, i liked your suggestions in the other thread so i thought id just add it to the schema as you suggested

> The schema, since it uses arrays does not allow for the omission of values however one could use `0` or `null` to signal no data for a particular categorization.

Reviewer: Just want to check the meaning of "one could use" here -- is the logic to accept `0` or `null` elements baked into the Valhalla code already? If so, does it switch to the speeds from `default_speeds.json`?

Complete side note, but I think it'll be interesting what criteria we use to deem a specific categorization "has too little info for us to generate an estimate, so we'll put a `null` here" when we're doing the final data aggregation step in conflation.

Author: >  is the logic to accept 0 or null elements baked into the Valhalla code already

nope. but @nilsnolde suggested it should be included in the spec and i agree. its important to remember that although valhalla is the first project to support these defaults the project should be able to be used by any project that wants to. so yeah in that way valhalla isnt up to spec yet, but we'll get it there!

> as too little info for us to generate an estimate

thats right, if we dont have enough observations of a certain type then we'll potentially want to put a `0` or `null` in there. what the threshold is i'm not sure yet :smile:

Reviewer: > valhalla is the first project to support these defaults the project should be able to be used by any project that wants to

Quick devil's advocate followup here, prefaced by the fact that I'm no expert 😅 . I think then there might be questions around how the projects that consume these generated configs will handle `null`s. E.g. if I'm conflation and I know that the defaults speeds are wildly inaccurate for `null` settings, then I would probably be lenient with any "minimum-data" thresholds and try to give an estimate for every specification. On the flip side if the default speeds are somehow amazing, then I might be strict and only make estimates when I have a ton of data.

But the issue is that conflation won't necessarily know how good the consuming projects handle `null`s. I could be wrong, but I also feel like it might be a little difficult for a consuming project (focused on routing) to even generate default speeds -- again I'm no expert but only hard coded values or some speed interpolation really come to mind.

Therefore maybe it makes sense to have conflation come up with these "default values", and just fill it in itself when there aren't enough datapoints to make a more pointed estimate? This way conflation is "forced" to do some fancier stuff, e.g. trying to extrapolate an estimate using datapoints from similar nearby regions, but this perhaps results in better ETAs than giving the job to the routing engine. Let me know if this makes sense!

Author: > little difficult for a consuming project (focused on routing) to even generate default speeds

that is a great question! there are 2 reasons why i think this is ok or preferred:

1. routing projects (like OSRM, Valhalla, GraphHopper) all existed before this OpenStreetMapSpeeds project which means they are already assigning speed values to edges in their respective graphs regardless of the difficulty. you just cannot build a router without solving this problem.
2. some folks will want to use their own data to build this json speed data for whatever region they have coverage in. by allowing missing data the person can pick and choose where the data applies and where it doesn't. its a "feature" :smile: 

because of these i think its completely reasonable for us to have spots in our suggested defaults where we can say you know what, we dont have data to make a decision so you are on your own. additionally resources for suggesting speeds already exist as some of the comments pointed out over there: https://github.com/valhalla/valhalla/issues/3021

> maybe it makes sense to have conflation come up with these "default values"

this isnt a bad idea at all. @nilsnolde suggested we use the OSM wiki suggestions for default speeds, to me that sounds like a good place to start for filler values but i do think that should be optional.

you're whole comment makes me think though that the schema should potentially be expanded to include metadata about the prevalence of observations that lead to the suggested speed. specifically maybe we should have both a speed and a count of observations which would allow the downstream consumer of this data make a choice, should i use this "low confidence" value or revert to my own built-in heuristic. if we did that we could always fill in default values (from where ever) but the user could see ah ha this particular type of edge had no or very few observations so i know the value is not statistically validated.

Reviewer: Yep, this all makes sense -- thanks for pointing out that routing projects _do_ in fact have preexisting default speed assignment algorithms. In that case it definitely makes sense to throw in `null`s.

But I had the exact same thought as you regarding including information on how confident our estimates are. Depending on whether we end up using statistical techniques, we could very easily include things like standard deviation or confidence intervals which bake in the `n` (# of measurements) used to arrive at our estimate. Or even simpler just include `n` as you mentioned or maybe even some sort of "0-100 confidence scale" that conflation defines.

This would enable us to give "poor confidence" estimates in areas with few / poor measurements, and therefore `null` can be reserved for cases where we have very very few or no measurements at all. It also enables each routing engine to choose what confidence levels to ignore depending on how good their default speed assignment algos are (and we've come full circle!).
